src/avl.py

```diff
@@ -0,0 +1,468 @@
+    def child(self, side):
+        """Returns the child on side (where side is "right" or "left").
+
+        Args:
+          side: One of "right" or "left". The side of the child we want to fetch."""
+
+        assert side in ("left", "right")
+        return getattr(self, "_" + side)
+
+    def set_child(self, side, new_node):
+        """Sets the child on side (where side is "right" or "left").
+
+        Args:
+          side: One of "right" or "left". The side of the child we want to set.
+          new_node: The node to set the child on side to.
+        """
+
+        assert side in ("left", "right")
+        setattr(self, "_%s" % side, new_node)
+        child = self.child(side)
+        child.parent = self
+        setattr(self, "%s_height" % side, child.max_height + 1 if child else 0)
+        self.fix_height_metadata()
+        if self.parent:
+            self.parent.fix_height_metadata()
+
+    def height(self, side):
+        """Returns the height of the child on side.
+
+        Args:
+          side: Which side the child we want the height of.
+        """
+
+        assert side in ("left", "right")
+        return getattr(self, "%s_height" % side)
+
+    def balance(self):
+        """Balances the tree if it is imbalanced."""
+
+        def values():
+            return set(node.value for node in self.container)
+
+        self.update_log("before balancing")
+
+        # if this assertion fails, the tree is more imbalanced than it ever should be.
+        assert self.balance_factor in (-2, -1, 0, 1, 2)
+
+        side = None
+        if self.balance_factor == 2:
+            side = "left"
+        if self.balance_factor == -2:
+            side = "right"
+
+        other_side = "left" if side == "right" else "right"
+
+        # this if is suspicious...
+        if side:
+            if self.child(side).child(other_side).value is not None \
+               and self.child(side).child(side).value is None:
+                self.child(side).rotate(side)
+
+            elif (self.child(side).child(other_side).value is not None
+                  and self.child(side).child(side).value is not None and
+                  # not sure why the following line matters, blindly portedofrom rust
+                  self.child(side).height(side) - self.child(side).height(other_side) < 0):
+
+                self.update_log("before rotate outside")
+                self.child(side).rotate(side)
+                self.update_log("after rotate outside", force=True)
+
+            self.rotate(other_side)
+
+        self.fix_height_metadata()
+        self.update_log("after balancing")
+
+        assert self.balance_factor in (-1, 0, 1)
+
+    def fix_height_metadata(self):
+        """Locally adjust the height metadata based on the height metadata of our children.
+
+        It is crucial that this code only be called when the children have sane height
+        metadata. If not, the whole tree will almost certainly break.
```

Comment:
Made [an issue](https://github.com/tehgeekmeister/dadabass/issues/10) that includes making this private.